sql/prometheus.sql

```diff
@@ -299,8 +299,7 @@ BEGIN
         EXECUTE format(
             $$
             CREATE VIEW %I AS 
-            SELECT prom_construct(m.time, l.metric_name, m.value, l.labels) AS sample,
-                   m.time AS time, l.metric_name AS name,  m.value AS value, l.labels AS labels
+            SELECT m.time AS time, l.metric_name AS name,  m.value AS value, l.labels AS labels
```

Comment:
Did you consider just creating an alternative view with a different name instead of changing the existing one. That might be safer for backwards compatibility. Otherwise you might have to make sure that you are running the right adapter. Or, does the adapter not query the raw samples? (In that case, why is this change necessary)?

If the adapter projected just `time, name, value, labels`, would `prom_construct` still have been called (thus making the query slower)?

Reply:
The adapter is not projecting `prom_construct` (sample) so this change is compatible with any adapter.  Here is an example to demonstrate the difference:

```
postgres=# \d+ metrics
                                    View "public.metrics"
 Column |           Type           | Collation | Nullable | Default | Storage  | Description
--------+--------------------------+-----------+----------+---------+----------+-------------
 sample | prom_sample              |           |          |         | plain    |
 time   | timestamp with time zone |           |          |         | plain    |
 name   | text                     |           |          |         | extended |
 value  | double precision         |           |          |         | plain    |
 labels | jsonb                    |           |          |         | extended |
View definition:
 SELECT prom_construct(m."time", l.metric_name, m.value, l.labels) AS sample,
    m."time",
    l.metric_name AS name,
    m.value,
    l.labels
   FROM metrics_values m
     JOIN metrics_labels l ON m.labels_id = l.id;
```
The view without `prom_construct`:
```
postgres=# \d+ metrics_2
                                   View "public.metrics_2"
 Column |           Type           | Collation | Nullable | Default | Storage  | Description
--------+--------------------------+-----------+----------+---------+----------+-------------
 time   | timestamp with time zone |           |          |         | plain    |
 name   | text                     |           |          |         | extended |
 value  | double precision         |           |          |         | plain    |
 labels | jsonb                    |           |          |         | extended |
View definition:
 SELECT m."time",
    l.metric_name AS name,
    m.value,
    l.labels
   FROM metrics_values m
     JOIN metrics_labels l ON m.labels_id = l.id;
```
Query execution for `metrics` view:

 ```
postgres=#  EXPLAIN ANALYZE SELECT * FROM metrics WHERE name = 'prometheus_engine_query_duration_seconds' AND time >= '2018-08-23T09:52:52Z' AND time <= '2018-08-30T22:03:52Z';
                                                                                     QUERY PLAN
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Nested Loop  (cost=0.28..1559751.07 rows=1137042 width=201) (actual time=13.441..4291.661 rows=1713489 loops=1)
   ->  Index Scan using metrics_labels_pkey on metrics_labels l  (cost=0.28..53.07 rows=18 width=157) (actual time=0.031..0.306 rows=18 loops=1)
         Filter: (metric_name = 'prometheus_engine_query_duration_seconds'::text)
         Rows Removed by Filter: 1104
   ->  Append  (cost=0.00..84938.47 rows=155350 width=20) (actual time=9.530..169.976 rows=95194 loops=18)
         ->  Seq Scan on metrics_values m  (cost=0.00..0.00 rows=1 width=20) (actual time=0.000..0.000 rows=0 loops=18)
               Filter: (("time" >= '2018-08-23 09:52:52+00'::timestamp with time zone) AND ("time" <= '2018-08-30 22:03:52+00'::timestamp with time zone) AND (l.id = labels_id))
         ->  Index Scan using _hyper_1_1_chunk_metrics_values_time_idx on _hyper_1_1_chunk m_1  (cost=0.56..2.79 rows=1 width=20) (actual time=0.007..0.007 rows=0 loops=18)
               Index Cond: (("time" >= '2018-08-23 09:52:52+00'::timestamp with time zone) AND ("time" <= '2018-08-30 22:03:52+00'::timestamp with time zone))
               Filter: (l.id = labels_id)
         ->  Bitmap Heap Scan on _hyper_1_4_chunk m_2  (cost=376.73..17968.92 rows=35039 width=20) (actual time=6.677..61.128 rows=23051 loops=18)
               Recheck Cond: (labels_id = l.id)
               Filter: (("time" >= '2018-08-23 09:52:52+00'::timestamp with time zone) AND ("time" <= '2018-08-30 22:03:52+00'::timestamp with time zone))
               Heap Blocks: exact=373533
               ->  Bitmap Index Scan on _hyper_1_4_chunk_metrics_values_lablels_id_only_idx  (cost=0.00..367.97 rows=35039 width=0) (actual time=3.610..3.610 rows=23051 loops=18)
                     Index Cond: (labels_id = l.id)
         ->  Bitmap Heap Scan on _hyper_1_2_chunk m_3  (cost=1143.04..60323.40 rows=106657 width=20) (actual time=8.231..83.294 rows=63625 loops=18)
               Recheck Cond: (labels_id = l.id)
               Filter: (("time" >= '2018-08-23 09:52:52+00'::timestamp with time zone) AND ("time" <= '2018-08-30 22:03:52+00'::timestamp with time zone))
               Heap Blocks: exact=365703
               ->  Bitmap Index Scan on _hyper_1_2_chunk_metrics_values_lablels_id_only_idx  (cost=0.00..1116.38 rows=106657 width=0) (actual time=5.201..5.201 rows=63625 loops=18)
                     Index Cond: (labels_id = l.id)
         ->  Bitmap Heap Scan on _hyper_1_3_chunk m_4  (cost=147.24..6643.36 rows=13652 width=20) (actual time=2.664..18.312 rows=8518 loops=18)
               Recheck Cond: (labels_id = l.id)
               Filter: (("time" >= '2018-08-23 09:52:52+00'::timestamp with time zone) AND ("time" <= '2018-08-30 22:03:52+00'::timestamp with time zone))
               Heap Blocks: exact=144198
               ->  Bitmap Index Scan on _hyper_1_3_chunk_metrics_values_lablels_id_only_idx  (cost=0.00..143.83 rows=13652 width=0) (actual time=1.561..1.561 rows=8518 loops=18)
                     Index Cond: (labels_id = l.id)
 Planning time: 0.836 ms
 Execution time: 4354.459 ms
(30 rows)
```
Query execution for `metrics_2`:

```
postgres=#  EXPLAIN ANALYZE SELECT * FROM metrics_2 WHERE name = 'prometheus_engine_query_duration_seconds' AND time >= '2018-08-23T09:52:52Z' AND time <= '2018-08-30T22:03:52Z';
                                                                                     QUERY PLAN
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Nested Loop  (cost=0.28..1556908.47 rows=1137042 width=169) (actual time=13.748..3196.374 rows=1713489 loops=1)
   ->  Index Scan using metrics_labels_pkey on metrics_labels l  (cost=0.28..53.07 rows=18 width=157) (actual time=0.035..0.315 rows=18 loops=1)
         Filter: (metric_name = 'prometheus_engine_query_duration_seconds'::text)
         Rows Removed by Filter: 1104
   ->  Append  (cost=0.00..84938.47 rows=155350 width=20) (actual time=9.548..167.674 rows=95194 loops=18)
         ->  Seq Scan on metrics_values m  (cost=0.00..0.00 rows=1 width=20) (actual time=0.000..0.000 rows=0 loops=18)
               Filter: (("time" >= '2018-08-23 09:52:52+00'::timestamp with time zone) AND ("time" <= '2018-08-30 22:03:52+00'::timestamp with time zone) AND (l.id = labels_id))
         ->  Index Scan using _hyper_1_1_chunk_metrics_values_time_idx on _hyper_1_1_chunk m_1  (cost=0.56..2.79 rows=1 width=20) (actual time=0.007..0.007 rows=0 loops=18)
               Index Cond: (("time" >= '2018-08-23 09:52:52+00'::timestamp with time zone) AND ("time" <= '2018-08-30 22:03:52+00'::timestamp with time zone))
               Filter: (l.id = labels_id)
         ->  Bitmap Heap Scan on _hyper_1_4_chunk m_2  (cost=376.73..17968.92 rows=35039 width=20) (actual time=6.721..60.441 rows=23051 loops=18)
               Recheck Cond: (labels_id = l.id)
               Filter: (("time" >= '2018-08-23 09:52:52+00'::timestamp with time zone) AND ("time" <= '2018-08-30 22:03:52+00'::timestamp with time zone))
               Heap Blocks: exact=373533
               ->  Bitmap Index Scan on _hyper_1_4_chunk_metrics_values_lablels_id_only_idx  (cost=0.00..367.97 rows=35039 width=0) (actual time=3.642..3.642 rows=23051 loops=18)
                     Index Cond: (labels_id = l.id)
         ->  Bitmap Heap Scan on _hyper_1_2_chunk m_3  (cost=1143.04..60323.40 rows=106657 width=20) (actual time=8.200..81.624 rows=63625 loops=18)
               Recheck Cond: (labels_id = l.id)
               Filter: (("time" >= '2018-08-23 09:52:52+00'::timestamp with time zone) AND ("time" <= '2018-08-30 22:03:52+00'::timestamp with time zone))
               Heap Blocks: exact=365703
               ->  Bitmap Index Scan on _hyper_1_2_chunk_metrics_values_lablels_id_only_idx  (cost=0.00..1116.38 rows=106657 width=0) (actual time=5.177..5.177 rows=63625 loops=18)
                     Index Cond: (labels_id = l.id)
         ->  Bitmap Heap Scan on _hyper_1_3_chunk m_4  (cost=147.24..6643.36 rows=13652 width=20) (actual time=2.638..18.171 rows=8518 loops=18)
               Recheck Cond: (labels_id = l.id)
               Filter: (("time" >= '2018-08-23 09:52:52+00'::timestamp with time zone) AND ("time" <= '2018-08-30 22:03:52+00'::timestamp with time zone))
               Heap Blocks: exact=144198
               ->  Bitmap Index Scan on _hyper_1_3_chunk_metrics_values_lablels_id_only_idx  (cost=0.00..143.83 rows=13652 width=0) (actual time=1.535..1.535 rows=8518 loops=18)
                     Index Cond: (labels_id = l.id)
 Planning time: 0.996 ms
 Execution time: 3259.103 ms
(30 rows)
```
Query execution seems identical. Can't tell if function is being called, but there is some performance hit.